File: backend-verification-service/src/services/database.service.ts
import { Pool, PoolClient } from 'pg';
import { Logger } from 'winston';

export interface Document {
  id: string;
  cid: string;
  title: string;
  mime_type: string;
  source_url?: string;
  content: string;
  chunks: string[];
  chunk_embeddings: number[][];
  created_at: Date;
}

export interface VerificationRequest {
  id: string;
  claim_text: string;
  confidence: number;
  status: 'VERIFIED' | 'UNVERIFIED' | 'UNKNOWN';
  doc_ids: string[];
  source?: string;
  extension_version?: string;
  processing_time_ms: number;
  created_at: Date;
}

export interface SearchOptions {
  limit: number;
  similarityThreshold: number;
}

export interface ApiKey {
  id: string;
  key: string;
  userId: string;
  organization: string;
  permissions: string[];
  tier: string;
  dailyQuota: number;
  monthlyQuota: number;
  rateLimit: number;
  isActive: boolean;
  expiresAt?: Date;
  createdAt: Date;
  lastUsedAt?: Date;
}

export interface ApiKeyUsage {
  id: string;
  apiKeyId: string;
  dailyUsage: number;
  monthlyUsage: number;
  usageDate: string;
  usageMonth: string;
}

export class DatabaseService {
  constructor(private pool: Pool, private logger: Logger) {}

  async initialize(): Promise<void> {
    try {
      // Test connection
      await this.pool.query('SELECT NOW()');
      this.logger.info('Database connection established');
      
      // Ensure pgvector extension is installed
      await this.pool.query('CREATE EXTENSION IF NOT EXISTS vector');
      
      // Create tables if they don't exist
      await this.createTables();
      
      this.logger.info('Database initialized successfully');
    } catch (error) {
      this.logger.error('Database initialization failed:', error);
      throw error;
    }
  }

  async healthCheck(): Promise<boolean> {
    try {
      await this.pool.query('SELECT 1');
      return true;
    } catch (error) {
      this.logger.error('Database health check failed:', error);
      return false;
    }
  }

  async storeDocument(document: Omit<Document, 'id' | 'created_at'>): Promise<string> {
    const client = await this.pool.connect();
    
    try {
      await client.query('BEGIN');
      
      // Insert document
      const docResult = await client.query(
        `INSERT INTO source_documents (cid, title, mime_type, source_url, content)
         VALUES ($1, $2, $3, $4, $5)
         RETURNING id`,
        [document.cid, document.title, document.mime_type, document.source_url, document.content]
      );
      
      const documentId = docResult.rows[0].id;
      
      // Store embeddings for each chunk
      for (let i = 0; i < document.chunks.length; i++) {
        const chunk = document.chunks[i];
        const embedding = document.chunk_embeddings[i];
        
        await client.query(
          `INSERT INTO document_chunks (document_id, chunk_index, content, embedding)
           VALUES ($1, $2, $3, $4)`,
          [documentId, i, chunk, embedding]
        );
      }
      
      await client.query('COMMIT');
      
      this.logger.info('Document stored successfully', { documentId, title: document.title });
      return documentId;
      
    } catch (error) {
      await client.query('ROLLBACK');
      this.logger.error('Failed to store document:', error);
      throw error;
    } finally {
      client.release();
    }
  }

  async searchSimilarDocuments(
    queryEmbedding: number[],
    options: SearchOptions
  ): Promise<Array<Document & { similarity: number }>> {
    try {
      const result = await this.pool.query(
        `SELECT 
           sd.id,
           sd.cid,
           sd.title,
           sd.mime_type,
           sd.source_url,
           sd.content,
           dc.content as chunk_content,
           dc.embedding <-> $1 as similarity
         FROM source_documents sd
         JOIN document_chunks dc ON sd.id = dc.document_id
         WHERE dc.embedding <-> $1 < $2
         ORDER BY similarity ASC
         LIMIT $3`,
        [queryEmbedding, 1 - options.similarityThreshold, options.limit]
      );
      
      // Group by document and get the best similarity score
      const documents = new Map<string, any>();
      
      for (const row of result.rows) {
        if (!documents.has(row.id)) {
          documents.set(row.id, {
            id: row.id,
            cid: row.cid,
            title: row.title,
            mime_type: row.mime_type,
            source_url: row.source_url,
            content: row.content,
            similarity: row.similarity
          });
        } else {
          // Keep the best (lowest) similarity score
          const existing = documents.get(row.id);
          if (row.similarity < existing.similarity) {
            existing.similarity = row.similarity;
          }
        }
      }
      
      return Array.from(documents.values()).sort((a, b) => a.similarity - b.similarity);
      
    } catch (error) {
      this.logger.error('Similarity search failed:', error);
      throw error;
    }
  }

  async storeVerificationRequest(request: VerificationRequest): Promise<void> {
    try {
      await this.pool.query(
        `INSERT INTO verification_requests 
         (id, claim_text, confidence, status, doc_ids, source, extension_version, processing_time_ms, created_at)
         VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9)`,
        [
          request.id,
          request.claim_text,
          request.confidence,
          request.status,
          request.doc_ids,
          request.source,
          request.extension_version,
          request.processing_time_ms,
          request.created_at
        ]
      );
      
      this.logger.debug('Verification request stored', { requestId: request.id });
      
    } catch (error) {
      this.logger.error('Failed to store verification request:', error);
      throw error;
    }
  }

  async getDocuments(page: number = 1, limit: number = 20): Promise<{
    documents: Document[];
    total: number;
    page: number;
    totalPages: number;
  }> {
    try {
      const offset = (page - 1) * limit;
      
      // Get total count
      const countResult = await this.pool.query('SELECT COUNT(*) FROM source_documents');
      const total = parseInt(countResult.rows[0].count);
      
      // Get documents
      const result = await this.pool.query(
        `SELECT id, cid, title, mime_type, source_url, content, created_at
         FROM source_documents
         ORDER BY created_at DESC
         LIMIT $1 OFFSET $2`,
        [limit, offset]
      );
      
      return {
        documents: result.rows.map(row => ({
          ...row,
          chunks: [],
          chunk_embeddings: []
        })),
        total,
        page,
        totalPages: Math.ceil(total / limit)
      };
      
    } catch (error) {
      this.logger.error('Failed to get documents:', error);
      throw error;
    }
  }

  async getVerificationAnalytics(
    startDate?: string,
    endDate?: string,
    source?: string
  ): Promise<{
    totalRequests: number;
    verifiedCount: number;
    unverifiedCount: number;
    unknownCount: number;
    averageConfidence: number;
    averageProcessingTime: number;
    requestsBySource: Record<string, number>;
    requestsByDate: Array<{ date: string; count: number }>;
  }> {
    try {
      let whereClause = 'WHERE 1=1';
      const params: any[] = [];
      let paramIndex = 1;
      
      if (startDate) {
        whereClause += ` AND created_at >= $${paramIndex++}`;
        params.push(startDate);
      }
      
      if (endDate) {
        whereClause += ` AND created_at <= $${paramIndex++}`;
        params.push(endDate);
      }
      
      if (source) {
        whereClause += ` AND source = $${paramIndex++}`;
        params.push(source);
      }
      
      // Get basic stats
      const statsResult = await this.pool.query(
        `SELECT 
           COUNT(*) as total_requests,
           COUNT(CASE WHEN status = 'VERIFIED' THEN 1 END) as verified_count,
           COUNT(CASE WHEN status = 'UNVERIFIED' THEN 1 END) as unverified_count,
           COUNT(CASE WHEN status = 'UNKNOWN' THEN 1 END) as unknown_count,
           AVG(confidence) as avg_confidence,
           AVG(processing_time_ms) as avg_processing_time
         FROM verification_requests
         ${whereClause}`,
        params
      );
      
      const stats = statsResult.rows[0];
      
      // Get requests by source
      const sourceResult = await this.pool.query(
        `SELECT source, COUNT(*) as count
         FROM verification_requests
         ${whereClause}
         GROUP BY source
         ORDER BY count DESC`,
        params
      );
      
      const requestsBySource: Record<string, number> = {};
      sourceResult.rows.forEach(row => {
        requestsBySource[row.source || 'unknown'] = parseInt(row.count);
      });
      
      // Get requests by date
      const dateResult = await this.pool.query(
        `SELECT 
           DATE(created_at) as date,
           COUNT(*) as count
         FROM verification_requests
         ${whereClause}
         GROUP BY DATE(created_at)
         ORDER BY date DESC
         LIMIT 30`,
        params
      );
      
      return {
        totalRequests: parseInt(stats.total_requests),
        verifiedCount: parseInt(stats.verified_count),
        unverifiedCount: parseInt(stats.unverified_count),
        unknownCount: parseInt(stats.unknown_count),
        averageConfidence: parseFloat(stats.avg_confidence) || 0,
        averageProcessingTime: parseFloat(stats.avg_processing_time) || 0,
        requestsBySource,
        requestsByDate: dateResult.rows.map(row => ({
          date: row.date,
          count: parseInt(row.count)
        }))
      };
      
    } catch (error) {
      this.logger.error('Failed to get verification analytics:', error);
      throw error;
    }
  }

  async getUniqueUsers(startDate?: string, endDate?: string): Promise<number> {
    try {
      let whereClause = 'WHERE 1=1';
      const params: any[] = [];
      let paramIndex = 1;
      
      if (startDate) {
        whereClause += ` AND created_at >= $${paramIndex++}`;
        params.push(startDate);
      }
      
      if (endDate) {
        whereClause += ` AND created_at <= $${paramIndex++}`;
        params.push(endDate);
      }
      
      const uniqueUsersQuery = `
        SELECT COUNT(DISTINCT 
          CASE 
            WHEN source = 'extension' THEN substring(claim_text FROM 1 FOR 8)
            ELSE source 
          END
        ) as unique_users
        FROM verification_requests 
        ${whereClause}
      `;
      
      const result = await this.pool.query(uniqueUsersQuery, params);
      return parseInt(result.rows[0]?.unique_users || '0');
      
    } catch (error) {
      this.logger.error('Failed to get unique users count:', error);
      throw error;
    }
  }

  async validateApiKey(apiKey: string): Promise<ApiKey | null> {
    try {
      const result = await this.pool.query(
        `SELECT ak.*, u.organization 
         FROM api_keys ak 
         JOIN users u ON ak.user_id = u.id 
         WHERE ak.key = $1 AND ak.is_active = true 
         AND (ak.expires_at IS NULL OR ak.expires_at > NOW())`,
        [apiKey]
      );
      
      if (result.rows.length === 0) {
        return null;
      }
      
      const row = result.rows[0];
      
      // Update last used timestamp
      await this.pool.query(
        'UPDATE api_keys SET last_used_at = NOW() WHERE id = $1',
        [row.id]
      );
      
      return {
        id: row.id,
        key: row.key,
        userId: row.user_id,
        organization: row.organization,
        permissions: row.permissions,
        tier: row.tier,
        dailyQuota: row.daily_quota,
        monthlyQuota: row.monthly_quota,
        rateLimit: row.rate_limit,
        isActive: row.is_active,
        expiresAt: row.expires_at,
        createdAt: row.created_at,
        lastUsedAt: row.last_used_at
      };
    } catch (error) {
      this.logger.error('Failed to validate API key:', error);
      throw error;
    }
  }

  async getApiKeyUsage(apiKeyId: string): Promise<{ daily: number; monthly: number }> {
    try {
      const now = new Date();
      const today = now.toISOString().split('T')[0];
      const thisMonth = now.toISOString().substring(0, 7);
      
      const result = await this.pool.query(
        `SELECT 
           COALESCE(SUM(CASE WHEN usage_date = $2 THEN daily_usage ELSE 0 END), 0) as daily_usage,
           COALESCE(SUM(CASE WHEN usage_month = $3 THEN monthly_usage ELSE 0 END), 0) as monthly_usage
         FROM api_key_usage 
         WHERE api_key_id = $1`,
        [apiKeyId, today, thisMonth]
      );
      
      return {
        daily: parseInt(result.rows[0]?.daily_usage || '0'),
        monthly: parseInt(result.rows[0]?.monthly_usage || '0')
      };
    } catch (error) {
      this.logger.error('Failed to get API key usage:', error);
      throw error;
    }
  }

  async incrementApiKeyUsage(apiKeyId: string): Promise<void> {
    try {
      const now = new Date();
      const today = now.toISOString().split('T')[0];
      const thisMonth = now.toISOString().substring(0, 7);
      
      await this.pool.query(
        `INSERT INTO api_key_usage (api_key_id, usage_date, usage_month, daily_usage, monthly_usage)
         VALUES ($1, $2, $3, 1, 1)
         ON CONFLICT (api_key_id, usage_date, usage_month)
         DO UPDATE SET 
           daily_usage = api_key_usage.daily_usage + 1,
           monthly_usage = api_key_usage.monthly_usage + 1`,
        [apiKeyId, today, thisMonth]
      );
    } catch (error) {
      this.logger.error('Failed to increment API key usage:', error);
      throw error;
    }
  }

  private async createTables(): Promise<void> {
    const client = await this.pool.connect();
    
    try {
      await client.query('BEGIN');
      
      // Create source_documents table
      await client.query(`
        CREATE TABLE IF NOT EXISTS source_documents (
          id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
          cid TEXT NOT NULL UNIQUE,
          title TEXT NOT NULL,
          mime_type TEXT NOT NULL,
          source_url TEXT,
          content TEXT NOT NULL,
          created_at TIMESTAMPTZ DEFAULT NOW()
        )
      `);
      
      // Create document_chunks table with vector support
      await client.query(`
        CREATE TABLE IF NOT EXISTS document_chunks (
          id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
          document_id UUID NOT NULL REFERENCES source_documents(id) ON DELETE CASCADE,
          chunk_index INTEGER NOT NULL,
          content TEXT NOT NULL,
          embedding vector(1536),
          created_at TIMESTAMPTZ DEFAULT NOW()
        )
      `);
      
      // Create verification_requests table
      await client.query(`
        CREATE TABLE IF NOT EXISTS verification_requests (
          id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
          claim_text TEXT NOT NULL,
          confidence NUMERIC,
          status VARCHAR(16) CHECK (status IN ('VERIFIED', 'UNVERIFIED', 'UNKNOWN')),
          doc_ids UUID[],
          source VARCHAR(50),
          extension_version VARCHAR(20),
          processing_time_ms INTEGER,
          created_at TIMESTAMPTZ DEFAULT NOW()
        )
      `);
      
      // Create users table
      await client.query(`
        CREATE TABLE IF NOT EXISTS users (
          id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
          email VARCHAR(255) UNIQUE NOT NULL,
          organization VARCHAR(255) NOT NULL,
          tier VARCHAR(50) DEFAULT 'free',
          is_active BOOLEAN DEFAULT true,
          created_at TIMESTAMPTZ DEFAULT NOW(),
          updated_at TIMESTAMPTZ DEFAULT NOW()
        )
      `);
      
      // Create api_keys table
      await client.query(`
        CREATE TABLE IF NOT EXISTS api_keys (
          id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
          key VARCHAR(255) UNIQUE NOT NULL,
          user_id UUID NOT NULL REFERENCES users(id) ON DELETE CASCADE,
          name VARCHAR(255) NOT NULL,
          permissions TEXT[] DEFAULT '{"read"}',
          tier VARCHAR(50) DEFAULT 'free',
          daily_quota INTEGER DEFAULT 1000,
          monthly_quota INTEGER DEFAULT 30000,
          rate_limit INTEGER DEFAULT 100,
          is_active BOOLEAN DEFAULT true,
          expires_at TIMESTAMPTZ,
          created_at TIMESTAMPTZ DEFAULT NOW(),
          last_used_at TIMESTAMPTZ
        )
      `);
      
      // Create api_key_usage table
      await client.query(`
        CREATE TABLE IF NOT EXISTS api_key_usage (
          id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
          api_key_id UUID NOT NULL REFERENCES api_keys(id) ON DELETE CASCADE,
          usage_date DATE NOT NULL,
          usage_month VARCHAR(7) NOT NULL,
          daily_usage INTEGER DEFAULT 0,
          monthly_usage INTEGER DEFAULT 0,
          created_at TIMESTAMPTZ DEFAULT NOW(),
          updated_at TIMESTAMPTZ DEFAULT NOW(),
          UNIQUE(api_key_id, usage_date, usage_month)
        )
      `);
      
      // Create indexes
      await client.query(`
        CREATE INDEX IF NOT EXISTS idx_document_chunks_embedding 
        ON document_chunks USING ivfflat (embedding vector_cosine_ops)
        WITH (lists = 100)
      `);
      
      await client.query(`
        CREATE INDEX IF NOT EXISTS idx_verification_requests_created_at 
        ON verification_requests (created_at)
      `);
      
      await client.query(`
        CREATE INDEX IF NOT EXISTS idx_verification_requests_status 
        ON verification_requests (status)
      `);
      
      await client.query(`
        CREATE INDEX IF NOT EXISTS idx_api_keys_key 
        ON api_keys (key)
      `);
      
      await client.query(`
        CREATE INDEX IF NOT EXISTS idx_api_keys_user_id 
        ON api_keys (user_id)
      `);
      
      await client.query(`
        CREATE INDEX IF NOT EXISTS idx_api_key_usage_api_key_id 
        ON api_key_usage (api_key_id)
      `);
      
      await client.query(`
        CREATE INDEX IF NOT EXISTS idx_api_key_usage_date 
        ON api_key_usage (usage_date)
      `);
      
      await client.query(`
        CREATE INDEX IF NOT EXISTS idx_users_email 
        ON users (email)
      `);
      
      await client.query('COMMIT');
      
    } catch (error) {
      await client.query('ROLLBACK');
      throw error;
    } finally {
      client.release();
    }
  }
} 

File: data-pipeline/src/veritas_data_pipeline/services/database_service.py
"""
Database Service for Veritas Data Pipeline
Handles PostgreSQL operations and vector storage
"""
import os
import uuid
import structlog
from typing import List, Dict, Any, Optional
import asyncpg

logger = structlog.get_logger()

class DatabaseService:
    """Database service for PostgreSQL operations"""
    
    def __init__(self):
        self.pool: Optional[asyncpg.Pool] = None
        self.connection_string = os.getenv('DATABASE_URL')
        if not self.connection_string:
            raise ValueError("DATABASE_URL environment variable is required")
    
    async def initialize(self) -> None:
        """Initialize database connection pool"""
        try:
            self.pool = await asyncpg.create_pool(
                self.connection_string,
                min_size=5,
                max_size=20,
                command_timeout=60
            )
            
            # Test connection
            async with self.pool.acquire() as conn:
                await conn.execute('SELECT 1')
            
            # Ensure tables exist
            await self.create_tables()
            
            logger.info("Database service initialized successfully")
            
        except Exception as e:
            logger.error("Database service initialization failed", error=str(e))
            raise
    
    async def create_tables(self) -> None:
        """Create database tables if they don't exist"""
        try:
            async with self.pool.acquire() as conn:
                # Create source_documents table
                await conn.execute('''
                    CREATE TABLE IF NOT EXISTS source_documents (
                        id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
                        cid TEXT NOT NULL UNIQUE,
                        title TEXT NOT NULL,
                        mime_type TEXT NOT NULL,
                        source_url TEXT,
                        content TEXT,
                        created_at TIMESTAMPTZ DEFAULT NOW(),
                        updated_at TIMESTAMPTZ DEFAULT NOW()
                    )
                ''')
                
                # Create document_chunks table
                await conn.execute('''
                    CREATE TABLE IF NOT EXISTS document_chunks (
                        id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
                        document_id UUID REFERENCES source_documents(id) ON DELETE CASCADE,
                        chunk_index INTEGER NOT NULL,
                        content TEXT NOT NULL,
                        embedding VECTOR(384),
                        created_at TIMESTAMPTZ DEFAULT NOW()
                    )
                ''')
                
                # Create verification_requests table
                await conn.execute('''
                    CREATE TABLE IF NOT EXISTS verification_requests (
                        id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
                        claim_text TEXT NOT NULL,
                        confidence NUMERIC(3,2),
                        status TEXT NOT NULL,
                        doc_ids UUID[],
                        source TEXT,
                        extension_version TEXT,
                        processing_time_ms INTEGER,
                        created_at TIMESTAMPTZ DEFAULT NOW()
                    )
                ''')
                
                # Create indexes
                await conn.execute('CREATE INDEX IF NOT EXISTS idx_source_documents_cid ON source_documents(cid)')
                await conn.execute('CREATE INDEX IF NOT EXISTS idx_document_chunks_document_id ON document_chunks(document_id)')
                await conn.execute('CREATE INDEX IF NOT EXISTS idx_verification_requests_created_at ON verification_requests(created_at)')
                
                # Enable vector extension if not already enabled
                await conn.execute('CREATE EXTENSION IF NOT EXISTS vector')
                
            logger.info("Database tables created successfully")
            
        except Exception as e:
            logger.error("Failed to create database tables", error=str(e))
            raise
    
    async def health_check(self) -> bool:
        """Check database health"""
        try:
            async with self.pool.acquire() as conn:
                await conn.execute('SELECT 1')
            return True
        except Exception as e:
            logger.error("Database health check failed", error=str(e))
            return False
    
    def generate_document_id(self) -> str:
        """Generate a unique document ID"""
        return str(uuid.uuid4())
    
    def generate_batch_id(self) -> str:
        """Generate a unique batch ID"""
        return str(uuid.uuid4())
    
    async def create_document_record(self, document_id: str, filename: str, status: str) -> None:
        """Create initial document record"""
        try:
            async with self.pool.acquire() as conn:
                await conn.execute('''
                    INSERT INTO source_documents (id, title, mime_type, status)
                    VALUES ($1, $2, $3, $4)
                ''', document_id, filename, 'application/octet-stream', status)
            
            logger.info("Document record created", document_id=document_id, filename=filename)
            
        except Exception as e:
            logger.error("Failed to create document record", 
                        document_id=document_id, error=str(e))
            raise
    
    async def store_document(self, document_data: Dict[str, Any]) -> str:
        """Store document and its chunks in database"""
        try:
            document_id = str(uuid.uuid4())
            
            async with self.pool.acquire() as conn:
                async with conn.transaction():
                    # Insert document
                    await conn.execute('''
                        INSERT INTO source_documents (id, cid, title, mime_type, source_url, content)
                        VALUES ($1, $2, $3, $4, $5, $6)
                    ''', document_id, document_data['cid'], document_data['title'],
                        document_data['mime_type'], document_data.get('source_url', ''),
                        document_data.get('content', ''))
                    
                    # Insert chunks with embeddings
                    chunks = document_data.get('chunks', [])
                    chunk_embeddings = document_data.get('chunk_embeddings', [])
                    
                    for i, (chunk, embedding) in enumerate(zip(chunks, chunk_embeddings)):
                        await conn.execute('''
                            INSERT INTO document_chunks (document_id, chunk_index, content, embedding)
                            VALUES ($1, $2, $3, $4)
                        ''', document_id, i, chunk, embedding)
            
            logger.info("Document stored successfully", 
                       document_id=document_id,
                       chunks_count=len(chunks))
            
            return document_id
            
        except Exception as e:
            logger.error("Failed to store document", error=str(e))
            raise
    
    async def search_similar_documents(self, embedding: List[float], 
                                     options: Dict[str, Any] = None) -> List[Dict[str, Any]]:
        """Search for similar documents using vector similarity"""
        try:
            limit = options.get('limit', 10) if options else 10
            similarity_threshold = options.get('similarityThreshold', 0.3) if options else 0.3
            
            async with self.pool.acquire() as conn:
                # Convert embedding to PostgreSQL vector format
                embedding_vector = f"[{','.join(map(str, embedding))}]"
                
                # Search using cosine similarity
                rows = await conn.fetch('''
                    SELECT 
                        dc.document_id,
                        dc.content,
                        dc.chunk_index,
                        sd.title,
                        sd.cid,
                        1 - (dc.embedding <=> $1) as similarity
                    FROM document_chunks dc
                    JOIN source_documents sd ON dc.document_id = sd.id
                    WHERE 1 - (dc.embedding <=> $1) > $2
                    ORDER BY dc.embedding <=> $1
                    LIMIT $3
                ''', embedding_vector, similarity_threshold, limit)
                
                documents = []
                for row in rows:
                    documents.append({
                        'id': str(row['document_id']),
                        'content': row['content'],
                        'chunk_index': row['chunk_index'],
                        'title': row['title'],
                        'cid': row['cid'],
                        'similarity': float(row['similarity'])
                    })
                
                logger.info("Similar documents found", 
                           query_embedding_length=len(embedding),
                           documents_found=len(documents),
                           top_similarity=documents[0]['similarity'] if documents else 0.0)
                
                return documents
                
        except Exception as e:
            logger.error("Failed to search similar documents", error=str(e))
            return []
    
    async def store_verification_request(self, request_data: Dict[str, Any]) -> None:
        """Store verification request for analytics"""
        try:
            async with self.pool.acquire() as conn:
                await conn.execute('''
                    INSERT INTO verification_requests 
                    (id, claim_text, confidence, status, doc_ids, source, extension_version, processing_time_ms, created_at)
                    VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9)
                ''', request_data['id'], request_data['claim_text'], request_data['confidence'],
                    request_data['status'], request_data['doc_ids'], request_data.get('source'),
                    request_data.get('extension_version'), request_data.get('processing_time_ms'),
                    request_data.get('created_at'))
            
            logger.info("Verification request stored", request_id=request_data['id'])
            
        except Exception as e:
            logger.error("Failed to store verification request", error=str(e))
            # Don't raise error for analytics data
    
    async def get_documents(self, page: int = 1, limit: int = 20) -> Dict[str, Any]:
        """Get paginated list of documents"""
        try:
            offset = (page - 1) * limit
            
            async with self.pool.acquire() as conn:
                # Get total count
                total_count = await conn.fetchval('SELECT COUNT(*) FROM source_documents')
                
                # Get documents
                rows = await conn.fetch('''
                    SELECT id, cid, title, mime_type, source_url, created_at
                    FROM source_documents
                    ORDER BY created_at DESC
                    LIMIT $1 OFFSET $2
                ''', limit, offset)
                
                documents = []
                for row in rows:
                    documents.append({
                        'id': str(row['id']),
                        'cid': row['cid'],
                        'title': row['title'],
                        'mime_type': row['mime_type'],
                        'source_url': row['source_url'],
                        'created_at': row['created_at'].isoformat()
                    })
                
                return {
                    'documents': documents,
                    'total_count': total_count,
                    'page': page,
                    'limit': limit,
                    'total_pages': (total_count + limit - 1) // limit
                }
                
        except Exception as e:
            logger.error("Failed to get documents", error=str(e))
            return {'documents': [], 'total_count': 0, 'page': page, 'limit': limit, 'total_pages': 0}
    
    async def get_verification_analytics(self, start_date: str = None, 
                                       end_date: str = None, source: str = None) -> Dict[str, Any]:
        """Get verification analytics"""
        try:
            async with self.pool.acquire() as conn:
                # Build query conditions
                conditions = []
                params = []
                param_count = 0
                
                if start_date:
                    param_count += 1
                    conditions.append(f"created_at >= ${param_count}")
                    params.append(start_date)
                
                if end_date:
                    param_count += 1
                    conditions.append(f"created_at <= ${param_count}")
                    params.append(end_date)
                
                if source:
                    param_count += 1
                    conditions.append(f"source = ${param_count}")
                    params.append(source)
                
                where_clause = " AND ".join(conditions) if conditions else "1=1"
                
                # Get analytics
                analytics = await conn.fetchrow(f'''
                    SELECT 
                        COUNT(*) as total_requests,
                        AVG(confidence) as avg_confidence,
                        AVG(processing_time_ms) as avg_processing_time,
                        COUNT(CASE WHEN status = 'VERIFIED' THEN 1 END) as verified_count,
                        COUNT(CASE WHEN status = 'UNVERIFIED' THEN 1 END) as unverified_count,
                        COUNT(CASE WHEN status = 'UNKNOWN' THEN 1 END) as unknown_count
                    FROM verification_requests
                    WHERE {where_clause}
                ''', *params)
                
                return {
                    'total_requests': analytics['total_requests'],
                    'avg_confidence': float(analytics['avg_confidence']) if analytics['avg_confidence'] else 0.0,
                    'avg_processing_time': float(analytics['avg_processing_time']) if analytics['avg_processing_time'] else 0.0,
                    'verified_count': analytics['verified_count'],
                    'unverified_count': analytics['unverified_count'],
                    'unknown_count': analytics['unknown_count'],
                    'verification_rate': float(analytics['verified_count']) / float(analytics['total_requests']) if analytics['total_requests'] > 0 else 0.0
                }
                
        except Exception as e:
            logger.error("Failed to get verification analytics", error=str(e))
            return {}
    
    async def close(self) -> None:
        """Close database connection pool"""
        if self.pool:
            await self.pool.close()
            logger.info("Database connection pool closed") 